### ISSUE TITLE
Error: popen() failed

### ISSUE THREAD

**Issue by gurunath-p** [marian-nmt/marian#306]
- Installed marian as described in the official documentation from master(marian-dev). Version: v1.8.30
 - Was training a transformer model for BEA Shared task as given in this [script](https://github.com/grammatical/magec-wnut2019/blob/master/training/en/train.sh)
 - Training data consists of 100 million sentences.
 - When training on a machine with 64GB RAM and 4 2080 GPUs, the process occupied about ~55GB of RAM and around 10GB on each GPU. 
 - The process calls a validation script named validate.sh every 'n' iterations(Here n=1) which returns a score for the validation set. I encountered the following error during the validation phase:
```
[2019-12-18 11:13:08] Ep. 1 : Up. 1 : Sen. 3,680 : Cost 147.10514832 : Time 60.33s : 826.70 words/s : L.r. 1.8750e-08
[2019-12-18 11:13:11] Saving model weights and runtime parameters to ./model.npz.best-cross-entropy.npz
[2019-12-18 11:13:12] [valid] Ep. 1 : Up. 1 : cross-entropy : 240.344 : new best
[2019-12-18 11:14:39] Error: popen() failed!
[2019-12-18 11:14:39] Error: Aborted from std::__cxx11::string marian::utils::exec(const string&, const std::vector<std::__cxx11::basic_string<char> >&, const string&) in pretraining-bea2019/systems/tools/marian-dev/src/common/utils.cpp:126

[CALL STACK]
[0x5630499039b3]    marian::utils::  exec  (std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>> const&,  std::vector<std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>>,std::allocator<std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>>>> const&,  std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>> const&) + 0x593
[0x563049ccf559]    marian::TranslationValidator::  validate  (std::vector<std::shared_ptr<marian::ExpressionGraph>,std::allocator<std::shared_ptr<marian::ExpressionGraph>>> const&,  std::shared_ptr<marian::TrainingState const>) + 0x1bb9
[0x563049c529e5]    marian::Scheduler::  validate  (std::vector<std::shared_ptr<marian::ExpressionGraph>,std::allocator<std::shared_ptr<marian::ExpressionGraph>>> const&,  bool) + 0x165
[0x563049c74402]    marian::SyncGraphGroup::  update  (std::vector<std::shared_ptr<marian::data::Batch>,std::allocator<std::shared_ptr<marian::data::Batch>>>,  unsigned long) + 0x912
[0x563049c75562]    marian::SyncGraphGroup::  update  (std::shared_ptr<marian::data::Batch>) + 0xfe2
[0x5630498dc242]    marian::Train<marian::SyncGraphGroup>::  run  ()   + 0x6a2
[0x5630497feaa0]    mainTrainer  (int,  char**)                        + 0x220
[0x5630497be2c5]    main                                               + 0x35
[0x7f7d59b33b97]    __libc_start_main                                  + 0xe7
[0x5630497fcdca]    _start                                             + 0x2a

./mod_train.sh: line 31: 115546 Aborted                 (core dumped) $MARIAN/marian --type transformer --model $MODELDIR/model.npz -d 0 1 2 3 --train-sets $DATA/train.src $DATA/train.trg --shuffle-in-ram --tempdir tmp --vocabs $DATA/vocab.{spm,spm} --tied-embeddings-all --max-length 150 --enc-depth 6 --dec-depth 6 --transformer-heads 8 --dropout-src 0.3 --dropout-trg 0.1 --transformer-dropout 0.3 --transformer-dropout-ffn 0.1 --transformer-dropout-attention 0.1 --exponential-smoothing --label-smoothing 0.1 --mini-batch-fit -w 9000 --mini-batch 200 --maxi-batch
200 --sync-sgd --optimizer-delay 4 --learn-rate 0.0003 --lr-warmup 16000 --lr-decay-inv-sqrt 16000 --lr-report --optimizer-params 0.9 0.98 1e-09 --clip-norm 0 --cost-type cross-entropy --valid-metrics cross-entropy translation perplexity --valid-sets $DATA/devset.{err,cor} --valid-translation-output $MODELDIR/devset.out --quiet-translation --valid-script-path $MODELDIR/validate.sh --valid-mini-batch 16 --beam-size 12 --normalize 1.0 --early-stopping 5 --after-epochs 5 --valid-freq 1 --save-freq 5000 --disp-freq 500 --disp-first 5 --overwrite --keep-best --log $MODELDIR/train.log --valid-log $MODELDIR/valid.log
```
 -  Trying the above commands with about half the data i.e., 50 million sentences runs without any problem and occupies ~28-29 GB of RAM. 
 - Interestingly, removing that particular validation step named "translation" in the `--valid-metrics` options makes the training run without any hiccups.
 - Also tried setting ulimit(No. of file descriptors) to 20000 but in vain.
 - Tried varying the reserved workspace memory and batch-size, etc..doesn't seem to make any difference.
 - Is there something I am missing? Is this probably caused due to insufficient RAM? If it is, then is there a way to load sentences lazily or in parts?
Thanks in advance.

**Comment by tomsbergmanis** [comment]
@gurunath-p I encountered a similar problem. Did you figure out what went wrong?

**Comment by gurunath-p** [comment]
@tomsbergmanis I couldn't find out the exact reason...FWIW it may have been an issue with the amount of RAM because running the training with 70 million sentences(instead of the original 100 million sentences) didn't stop the training. Please let me know incase you find the reason.

**Comment by tomsbergmanis** [comment]
In my case, it seems to be because of a co-worker who had used up 90% or
RAM and I hadn't noticed it.


Toms Bergmanis

PhD Candidate at University of Edinburgh

http://homepages.inf.ed.ac.uk/s1044253

**Comment by emjotde** [comment]
So, I guess RAM or out of disk space issues? If you have more problems, feel free to re-open.
